---
title: "July 2024 Cas9 Production"
date: "July 11th, 2024"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=65), tidy=TRUE)
library(readxl)
library(ggplot2)
library(ggpubr)
library(dplyr)
library(jtools)
library(formatR)

setwd("/home/helrick/Projects/sickkids/cas9-production")
dataset = read_excel("Clean-IMPC_Cas9_2021_03_annotated.xlsx")
```

## Preparing Data 

### Columns of Interest
First, subset the data to only the columns of interest 
```{r subset}
variables <- c("#G0 deletion event detected",
               "Founder Rate (per Embryos Transferred)",
               "Cellular Essential",
               "CpGsites",
               "pLI of Orthologs",
               "oe of Orthologs",
               "Has Omim Annotation",
               "Relative Chromosomal Position",
               "Staining Overlap",
               "AverageH3K27me3_Intron_PeakScore",
               "AverageH3K27ac_PeakScore",
               "all_stages_PercentileRank")
ds <- dataset[variables]
num_rows <- nrow(ds)
```

### Remove NAs and Inconsistencies

Initially there are `r num_rows` rows in the dataset. For analysis, remove rows with NAs in any of the following:

* Cellular Essential
* pLI of Human Orthologs
* oe of Human Orthologs
* Staining overlap (Giemsa positive)
* Embryonic expression

Or experiments that are both Methylated and Acetylated.
```{r cleanup}
# remove rows with NA values in the following fields:
ds <- ds[complete.cases(ds$`Cellular Essential`),]
ds <- ds[complete.cases(ds$`pLI of Orthologs`),]
ds <- ds[complete.cases(ds$`oe of Orthologs`),]
ds <- ds[complete.cases(ds$`Staining Overlap`),]
ds <- ds[complete.cases(ds$`all_stages_PercentileRank`),]

# remove experiments that are both Methylated and Acetylated
ds$acetylation <- as.numeric(ds$AverageH3K27ac_PeakScore)
ds$methylation <- as.numeric(ds$AverageH3K27me3_Intron_PeakScore)
ds$acetylation[is.na(ds$acetylation)] <- 0
ds$methylation[is.na(ds$methylation)] <- 0
ds <- subset(ds, !(ds$acetylation != 0 & ds$methylation != 0))

num_rows <- nrow(ds) # after removing NAs
```

After cleanup, there are `r num_rows` rows in the dataset. 

### Convert Variables to Numeric and Factor

Next, the variables need to be converted to numeric or factor variables for input to the models.

```{r convert}
# output of model, founders indicates success
Founders <- ifelse(ds$`#G0 deletion event detected` > 0, 1, 0)

# essentiality as binary
Essential <- as.factor(ds$`Cellular Essential`)
Essential <- ifelse(ds$`Cellular Essential` == "Essential", 1, 0)
# histone modifications as binary variables 
Acetylated = ifelse(ds$acetylation > 0, 1, 0)
Methylated = ifelse(ds$methylation > 0, 1, 0)
# staining as binary variable (gneg is 0, everything else is 1) 
GiemsaPositiveStain <- ifelse(ds$`Staining Overlap` != 'gneg', 1, 0)
# 'has OMIM annotation' as binary
OmimAnnotation <- as.factor(ds$`Has Omim Annotation`)
OmimAnnotation <- ifelse(ds$`Has Omim Annotation` == "t", 0, 1)

# remaining to numeric
EmbryonicExpression <- as.numeric(ds$all_stages_PercentileRank)
pLI <- as.numeric(ds$`pLI of Orthologs`)
oe <- as.numeric(ds$`oe of Orthologs`)
RelativeChromosomalLocation <- as.numeric(ds$`Relative Chromosomal Position`)
```

## Generate Models

Create general linear models with and without essentiality.
```{r models}
without_essentiality <- glm(Founders ~ EmbryonicExpression+pLI+oe+RelativeChromosomalLocation+Acetylated+Methylated+GiemsaPositiveStain+OmimAnnotation, data=ds)
with_essentiality <- glm(Founders ~ Essential+EmbryonicExpression+pLI+oe+RelativeChromosomalLocation+Acetylated+Methylated+GiemsaPositiveStain+OmimAnnotation, data=ds)
```

### GLM Statistics: Without Essentiality

Model Statistics:
```{r print_models_without, echo=FALSE}
summary_without_essentiality <- summary(without_essentiality)
print(summary_without_essentiality,digits=6)
```
Log Odds:
```{r, echo=FALSE}
stack(exp(without_essentiality$coefficients))
```

### GLM Statistics: With Essentiality
Model Statistics:
```{r print_models_with, echo=FALSE}
summary_with_essentiality <- summary(with_essentiality)
print(summary_with_essentiality,digits=6)
```
Log Odds:
```{r, echo=FALSE}
stack(exp(with_essentiality$coefficients))
```

## Plot Model Comparison
```{r plots, echo=FALSE}

without_essentiality_frame <- data.frame(Variable = rownames(summary(without_essentiality)$coef)[2:length(without_essentiality$coef)],
                                         Odds = exp(summary(without_essentiality)$coef[, 1][2:length(without_essentiality$coef)]),
                                         SE = summary(without_essentiality)$coef[, 2][2:length(without_essentiality$coef)],
                                         Model = "Without Essentiality")

with_essentiality_frame <- data.frame(Variable = rownames(summary(with_essentiality)$coef)[2:length(with_essentiality$coef)],
                                         Odds = exp(summary(with_essentiality)$coef[, 1][2:length(with_essentiality$coef)]),
                                         SE = summary(with_essentiality)$coef[, 2][2:length(with_essentiality$coef)],
                                         Model = "With Essentiality")

# reorder factor levels
without_essentiality_frame$Variable <- factor(without_essentiality_frame$Variable, 
                                              levels = c("EmbryonicExpression", "pLI", "oe", "Methylated", "Acetylated", "GiemsaPositiveStain", "RelativeChromosomalLocation", "OmimAnnotation"))
with_essentiality_frame$Variable <- factor(with_essentiality_frame$Variable, 
                                              levels = c("Essential", "EmbryonicExpression", "pLI", "oe", "Methylated", "Acetylated", "GiemsaPositiveStain", "RelativeChromosomalLocation", "OmimAnnotation"))

both_models <- data.frame(rbind(without_essentiality_frame, with_essentiality_frame))

interval1 <- -qnorm((1-0.9)/2)  # 90% multiplier
interval2 <- -qnorm((1-0.95)/2)  # 95% multiplier

zp1 <- ggplot(both_models, aes(colour = Model))
zp1 <- zp1 + geom_hline(yintercept = 1, colour = gray(1/2), lty = 2)
zp1 <- zp1 + geom_linerange(aes(x = Variable, ymin = Odds - SE*interval1,
                                ymax = Odds + SE*interval1),
                            lwd = 1, position = position_dodge(width = 1/2))
zp1 <- zp1 + geom_pointrange(aes(x = Variable, y = Odds, ymin = Odds - SE*interval2,
                                 ymax = Odds + SE*interval2),
                             lwd = 1/2, position = position_dodge(width = 1/2),
                             shape = 21, fill = "WHITE")
zp1 <- zp1 + coord_flip() + theme_bw() + scale_color_manual(values = c("#b0b0b0", "#545454")) + theme(axis.title.y = element_blank()) 

print(zp1)

png('GLM_2024-07-11.png')
plot(zp1)
dev.off() -> .
```

